[Cocoa] Add webm support to MediaRecorder

jyavenard opened the pull request:
#### d27040f0d729453dbf87696eddd077ec1f6d25d3
<pre>
[Cocoa] Add webm support to MediaRecorder
<a href="https://bugs.webkit.org/show_bug.cgi?id=281106">https://bugs.webkit.org/show_bug.cgi?id=281106</a>
<a href="https://rdar.apple.com/137560454">rdar://137560454</a>

Reviewed by Youenn Fablet.

We add a MediaRecorderPrivateWriterWebM that uses the mkvmuxer from libwebm and the WebCodecs video encoders.
The WebM MediaRecorder is behind a preferences, currently disabled by default.

The MediaRecorderPrivateWriterWebM is currently used in the GPU process, in place of the MediaRecorderPrivateWriterCocoa
for ease of integration.
Performing the encoding and muxing will be moved to the Web Process in a follow-up change.

Amending existing tests to exercise the code.

* Configurations/WebKitProjectPaths.xcconfig:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-AV-audio-video-dataavailable-gpuprocess-expected.txt:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-AV-audio-video-dataavailable-gpuprocess.html:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-expected.txt:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate.html:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-requestData-expected.txt:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-requestData.html:
* LayoutTests/http/wpt/mediarecorder/mimeType-expected.txt:
* LayoutTests/http/wpt/mediarecorder/mimeType.html:
* LayoutTests/http/wpt/mediarecorder/pause-recording-expected.txt:
* LayoutTests/http/wpt/mediarecorder/pause-recording.html:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-mimetype-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-peerconnection.https-expected.txt:
* LayoutTests/platform/glib/TestExpectations: GStreamer MediaRecorder always create a MP4 one.
* Source/ThirdParty/libwebrtc/CMakeLists.txt:
* Source/ThirdParty/libwebrtc/Configurations/BaseTarget-libwebm.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/libwebm.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.exp:
* Source/ThirdParty/libwebrtc/Source/third_party/libwebm/CMakeLists.txt:
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/Modules/mediarecorder/MediaRecorder.cpp:
(WebCore::MediaRecorder::isTypeSupported):
* Source/WebCore/Modules/mediarecorder/MediaRecorderProvider.cpp:
(WebCore::MediaRecorderProvider::isSupported):
(WebCore::MediaRecorderProvider::isWebMAndSupported):
* Source/WebCore/Modules/mediarecorder/MediaRecorderProvider.h:
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::span8):
(WebCore::createOpusPrivateData):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.cpp:
(WebCore::MediaRecorderPrivateAVFImpl::create):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.cpp: Copied from Source/WebCore/Modules/mediarecorder/MediaRecorderProvider.h.
(WebCore::MediaRecorderPrivateWriter::create):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h: Added.
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::create):
(WebCore::AudioSampleBufferCompressor::AudioSampleBufferCompressor):
(WebCore::AudioSampleBufferCompressor::initAudioConverterForSourceFormatDescription):
(WebCore::AudioSampleBufferCompressor::getOutputSampleBuffer const):
(WebCore::AudioSampleBufferCompressor::isEmpty const):
(WebCore::AudioSampleBufferCompressor::getOutputSampleBuffer): Deleted.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm:
(WebCore::MediaRecorderPrivateWriter::create):
(WebCore::MediaRecorderPrivateWriterAVFObjC::create):
(WebCore::MediaRecorderPrivateWriterAVFObjC::compressedVideoOutputBufferCallback):
(WebCore::MediaRecorderPrivateWriterAVFObjC::compressedAudioOutputBufferCallback):
(WebCore::MediaRecorderPrivateWriterAVFObjC::MediaRecorderPrivateWriterAVFObjC):
(WebCore::MediaRecorderPrivateWriterAVFObjC::~MediaRecorderPrivateWriterAVFObjC):
(WebCore::MediaRecorderPrivateWriterAVFObjC::close):
(WebCore::MediaRecorderPrivateWriterAVFObjC::initialize):
(WebCore::MediaRecorderPrivateWriterAVFObjC::processNewCompressedVideoSampleBuffers):
(WebCore::MediaRecorderPrivateWriterAVFObjC::processNewCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateWriterAVFObjC::startAssetWriter):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendCompressedAudioSampleBufferIfPossible):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendCompressedVideoSampleBufferIfPossible):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendCompressedVideoSampleBuffer):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendCompressedSampleBuffers):
(WebCore::MediaRecorderPrivateWriterAVFObjC::flushCompressedSampleBuffers):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendVideoFrame):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendAudioSampleBuffer):
(WebCore::MediaRecorderPrivateWriterAVFObjC::finishedFlushingSamples):
(WebCore::MediaRecorderPrivateWriterAVFObjC::stopRecording):
(WebCore::MediaRecorderPrivateWriterAVFObjC::fetchData):
(WebCore::MediaRecorderPrivateWriterAVFObjC::completeFetchData):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendData):
(WebCore::MediaRecorderPrivateWriterAVFObjC::takeData):
(WebCore::MediaRecorderPrivateWriterAVFObjC::pause):
(WebCore::MediaRecorderPrivateWriterAVFObjC::resume):
(WebCore::MediaRecorderPrivateWriterAVFObjC::mimeType const):
(WebCore::MediaRecorderPrivateWriterAVFObjC::audioBitRate const):
(WebCore::MediaRecorderPrivateWriterAVFObjC::videoBitRate const):
(WebCore::MediaRecorderPrivateWriter::compressedVideoOutputBufferCallback): Deleted.
(WebCore::MediaRecorderPrivateWriter::compressedAudioOutputBufferCallback): Deleted.
(WebCore::MediaRecorderPrivateWriter::MediaRecorderPrivateWriter): Deleted.
(WebCore::MediaRecorderPrivateWriter::~MediaRecorderPrivateWriter): Deleted.
(WebCore::MediaRecorderPrivateWriter::close): Deleted.
(WebCore::MediaRecorderPrivateWriter::initialize): Deleted.
(WebCore::MediaRecorderPrivateWriter::processNewCompressedVideoSampleBuffers): Deleted.
(WebCore::MediaRecorderPrivateWriter::processNewCompressedAudioSampleBuffers): Deleted.
(WebCore::MediaRecorderPrivateWriter::startAssetWriter): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendCompressedAudioSampleBufferIfPossible): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendCompressedVideoSampleBufferIfPossible): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendCompressedVideoSampleBuffer): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendCompressedSampleBuffers): Deleted.
(WebCore::MediaRecorderPrivateWriter::flushCompressedSampleBuffers): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendVideoFrame): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendAudioSampleBuffer): Deleted.
(WebCore::MediaRecorderPrivateWriter::finishedFlushingSamples): Deleted.
(WebCore::MediaRecorderPrivateWriter::stopRecording): Deleted.
(WebCore::MediaRecorderPrivateWriter::fetchData): Deleted.
(WebCore::MediaRecorderPrivateWriter::completeFetchData): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendData): Deleted.
(WebCore::MediaRecorderPrivateWriter::takeData): Deleted.
(WebCore::MediaRecorderPrivateWriter::pause): Deleted.
(WebCore::MediaRecorderPrivateWriter::resume): Deleted.
(WebCore::MediaRecorderPrivateWriter::mimeType const): Deleted.
(WebCore::MediaRecorderPrivateWriter::audioBitRate const): Deleted.
(WebCore::MediaRecorderPrivateWriter::videoBitRate const): Deleted.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.h: Copied from Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.h.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm: Added.
(WebCore::MediaRecorderPrivateWriterWebMDelegate::MediaRecorderPrivateWriterWebMDelegate):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addFrame):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::forceNewClusterOnNextFrame):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::finalize):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::hasAddedFrame const):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::protectedParent const):
(WebCore::MediaRecorderPrivateWriterWebM::create):
(WebCore::mkvCodeIcForMediaVideoCodecId):
(WebCore::codecStringForMediaVideoCodecId):
(WebCore::MediaRecorderPrivateWriterWebM::MediaRecorderPrivateWriterWebM):
(WebCore::MediaRecorderPrivateWriterWebM::~MediaRecorderPrivateWriterWebM):
(WebCore::MediaRecorderPrivateWriterWebM::close):
(WebCore::MediaRecorderPrivateWriterWebM::compressedAudioOutputBufferCallback):
(WebCore::MediaRecorderPrivateWriterWebM::initialize):
(WebCore::MediaRecorderPrivateWriterWebM::enqueueCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateWriterWebM::maybeStartWriting):
(WebCore::MediaRecorderPrivateWriterWebM::appendVideoFrame):
(WebCore::MediaRecorderPrivateWriterWebM::nextVideoFrameTime const):
(WebCore::MediaRecorderPrivateWriterWebM::resumeVideoTime const):
(WebCore::MediaRecorderPrivateWriterWebM::sampleTime const):
(WebCore::MediaRecorderPrivateWriterWebM::encodePendingVideoFrames):
(WebCore::MediaRecorderPrivateWriterWebM::appendAudioSampleBuffer):
(WebCore::MediaRecorderPrivateWriterWebM::flushEncodedQueues):
(WebCore::MediaRecorderPrivateWriterWebM::partiallyFlushEncodedQueues):
(WebCore::MediaRecorderPrivateWriterWebM::muxNextFrame):
(WebCore::MediaRecorderPrivateWriterWebM::stopRecording):
(WebCore::MediaRecorderPrivateWriterWebM::fetchData):
(WebCore::MediaRecorderPrivateWriterWebM::completeFetchData):
(WebCore::MediaRecorderPrivateWriterWebM::flushPendingData):
(WebCore::MediaRecorderPrivateWriterWebM::appendData):
(WebCore::MediaRecorderPrivateWriterWebM::takeData):
(WebCore::MediaRecorderPrivateWriterWebM::pause):
(WebCore::MediaRecorderPrivateWriterWebM::resume):
(WebCore::MediaRecorderPrivateWriterWebM::mimeType const):
(WebCore::MediaRecorderPrivateWriterWebM::audioBitRate const):
(WebCore::MediaRecorderPrivateWriterWebM::videoBitRate const):
(WebCore::MediaRecorderPrivateWriterWebM::maybeForceNewCluster):
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.cpp:
(WebKit::RemoteMediaRecorder::fetchData):
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.h:
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.messages.in:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionManagerProxy::seekableRangesVectorChanged):
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.cpp:
(WebKit::MediaRecorderPrivate::fetchData):

Canonical link: <a href="https://commits.webkit.org/285061@main">https://commits.webkit.org/285061@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9273c17bd814c462c526bc422e191b8fefaa4edc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71443 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50856 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24216 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75554 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22648 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73558 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58656 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22467 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/14893 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74509 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46158 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61541 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/36870 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42821 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19006 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20989 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/64568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64720 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19372 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77273 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/70692 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15677 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18544 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/64138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15720 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61581 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15793 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12284 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5918 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/92478 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46656 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/1435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20390 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47727 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49011 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47469 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/ThirdParty/libwebrtc/Configurations/libwebrtc.exp:
```diff
@@ -403,3 +403,29 @@ __ZN3rtc15CountIPMaskBitsERKNS_9IPAddressE
+__ZN6webrtc14RTPVideoHeaderC1Ev
+__ZN6webrtc14RTPVideoHeaderD1Ev
+__ZN6webrtc23VideoRtpDepacketizerVp815ParseRtpPayloadEN3rtc9ArrayViewIKhLln4711EEEPNS_14RTPVideoHeaderE
+__ZN6webrtc23VideoRtpDepacketizerVp915ParseRtpPayloadEN3rtc9ArrayViewIKhLln4711EEEPNS_14RTPVideoHeaderE
```
Why do we need these 4, they are not related to webm?

jyavenard replied:
I added all the ones that the linker complained about. It could be that the code got optimised differently with the new code in, I'm not sure.

youennf commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -1065,3 +1065,7 @@
 #define ENABLE_UNFAIR_LOCK 1
 #endif
+
+#if !defined(ENABLE_MEDIA_RECORDER_WEBM) && ENABLE(MEDIA_RECORDER) && ENABLE(OPUS) && ENABLE(WEB_CODECS)
```
Why WEB_CODECS?
It seems we should be using USE(LIWEBRTC) instead.

jyavenard replied:
VideoEncoder is only available with WEB_CODECS `#if ENABLE(WEB_CODECS)`

jyavenard replied:
I wanted to make the patch as small as possible and not have to also touch all those files VideoEncoder relies on

youennf replied:
Oh ok, sounds good. We should make VideoEncoder compiling without WEB_CODECS.

jyavenard replied:
I'll do this in a follow up patch

youennf commented on Source/WebCore/Modules/mediarecorder/MediaRecorderProvider.cpp:
```diff
@@ -69,23 +69,46 @@ bool MediaRecorderProvider::isSupported(const String& value)
+    if (equalLettersIgnoringASCIICase(containerType, "audio/mp4"_s) || equalLettersIgnoringASCIICase(containerType, "video/mp4"_s)) {
+        for (auto& codec : mimeType.codecs()) {
+            // FIXME: We should further validate parameters.
+            if (!startsWithLettersIgnoringASCIICase(codec, "avc1"_s) && !startsWithLettersIgnoringASCIICase(codec, "mp4a"_s))
```
This is sort of orthogonal to this patch. I guess this is fine, but then we should add some tests.
Or split it in its own patch.

jyavenard replied:
not really, this is required to determine if WebM is allowed or not and produce the proper error (throw)

jyavenard replied:
oh the change I made here is functionally identical to before, ContentType::codecs() was already doing `trim(isASCIIWhitespace<UChar>)` there was no need to do it a 2nd time

youennf commented on Source/WebCore/Modules/mediarecorder/MediaRecorderProvider.cpp:
```diff
@@ -69,23 +69,46 @@ bool MediaRecorderProvider::isSupported(const String& value)
+        }
+        return true;
+    }
+    return isWebMAndSupported(value);
```
```suggestion
    return isWebMAndOpusSupported(value);
```
or just `isWebMSupported `

jyavenard replied:
`WebMSupported` would return `true` if it's MP4 and supported which isn't what we want here.

We need to know both if it's WebM and if it's supported

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
```diff
@@ -44,18 +45,24 @@
 {
-    Ref compressor = adoptRef(*new AudioSampleBufferCompressor());
+    return create(kAudioFormatMPEG4AAC, callback, callbackObject);
+}
```
We should remove this one, or AudioFormatID format parameter optional with  kAudioFormatMPEG4AAC as the default value.

jyavenard replied:
this prevents having to create either a 2nd constructor or make a parameter optional.

youennf replied:
I would do:
`RefPtr<AudioSampleBufferCompressor> create(CMBufferQueueTriggerCallback, void* callbackObject, AudioFormatID = kAudioFormatMPEG4AAC)`, no need for this additional create.

jyavenard replied:
done

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.h:
```diff
@@ -64,28 +64,50 @@ struct MediaRecorderPrivateOptions;
 class WEBCORE_EXPORT MediaRecorderPrivateWriter : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<MediaRecorderPrivateWriter, WTF::DestructionThread::Main> {
```
Let's move MediaRecorderPrivateWriter interface to Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h, since it should no longer contain any COCOA specific construct.

jyavenard replied:
it's not used by GStreamer though.

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.h:
```diff
@@ -0,0 +1,154 @@
+    Seconds m_maxClusterDuration { 2_s };
+    MonotonicTime m_lastTimestamp;
+    Seconds m_lastMuxedSampleTime;
+    std::unique_ptr<MediaRecorderPrivateWriterWebMDelegate> m_delegate;
```
UniqueRef?

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
```diff
@@ -0,0 +1,601 @@
+        m_position += len;
+        return 0;
+    }
+    int64_t Position() const final { return m_position; }
```
Can these be private?

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
```diff
@@ -0,0 +1,601 @@
+
+    bool hasAddedFrame() const { return m_hasAddedFrame; }
+
+    Ref<MediaRecorderPrivateWriterWebM> protectedParent() const { return RefPtr { m_parent.get() }.releaseNonNull(); }
```
It is only used once and could be private as well apparently.

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
```diff
@@ -0,0 +1,601 @@
+static const char* mkvCodeIcForMediaVideoCodecId(FourCharCode codec)
+{
+    switch (codec) {
+    case 'vp08': return mkvmuxer::Tracks::kVp8CodecId;
```
```suggestion
    case 'vp08':
         return mkvmuxer::Tracks::kVp8CodecId;
```
Ditto below.

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
```diff
@@ -0,0 +1,601 @@
+        if (!asbd)
+            return;
+
+        m_audioTrackIndex = m_delegate->addAudioTrack(*asbd, mkvCodeIcForMediaVideoCodecId(m_audioCodec));
```
I would tend to pass m_audioCodec here and do the conversion at libwebm call site.
Ditto for video.

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
```diff
@@ -0,0 +1,601 @@
+    if (!m_audioFormatDescription) {
+        m_audioFormatDescription = PAL::CMSampleBufferGetFormatDescription(RefPtr { m_audioCompressor }->getOutputSampleBuffer());
+
+        const AudioStreamBasicDescription* const asbd = PAL::CMAudioFormatDescriptionGetStreamBasicDescription(m_audioFormatDescription.get());
```
```suggestion
        const auto* const asbd = PAL::CMAudioFormatDescriptionGetStreamBasicDescription(m_audioFormatDescription.get());
```

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
```diff
@@ -0,0 +1,601 @@
+            if (!result || !protectedThis)
+                return;
+            m_videoEncoder = result.value().moveToUniquePtr();
+            m_videoEncoder->setRates(m_videoBitsPerSecond, 0);
```
Should we call `encodePendingVideoFrames ` here?

jyavenard replied:
it is called, just after the call to `VideoEncoder::create(` ; but I can see how it might be preferable to do so here too.

youennf replied:
There is no m_videoEncoder yet after so encodePendingVideoFrames() below will be a no-op.

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
```diff
@@ -0,0 +1,601 @@
+
+    m_pendingVideoEncode++;
+
+    m_currentOperation = Ref { m_currentOperation }->whenSettled(MainThreadDispatcher::singleton(), [weakThis = ThreadSafeWeakPtr { *this }, this, pendingVideoFrame = std::exchange(m_pendingVideoFrames, { })]() mutable {
```
I would tend to only set m_currentOperation for flush operations.
Flush promise ensures that all encodes are done.

Something like:
if (m_currentOperation) {
   m_currentOperation->whenSettled(... { encodePendingVideoFrames(); })
   return;
}
while (! m_pendingVideoFrames.isEmpty())
   m_videoEncoder->encode(...);

jyavenard replied:
we want to encode the frames in the right order. So if new frames are coming in (while a flush is pending) we would output the current in-flight frame (they were the one available at the time of flush).

So we need to only start encoding the new frames once the previous ones have finished being encoded.

If we didn't append them to the current operation we would have frames out of order at the end.

jyavenard replied:
also with the change you proposed, we wouldn't get the video frames currently being encoded prior the flush returned

It's arguable that it would matter if we didn't ; but the spec of requestData does require that we return all the frames that we have received up to the time requestData  was called.

```
Let blob be the [Blob](https://w3c.github.io/FileAPI/#dfn-Blob) of collected data so far and let target be the [MediaRecorder](https://w3c.github.io/mediacapture-record/#mediarecorder) context object, then [fire a blob event](https://w3c.github.io/mediacapture-record/#fire-a-blob-event) named [dataavailable](https://w3c.github.io/mediacapture-record/#eventdef-mediarecorder-dataavailable) at target with blob. (Note that blob will be empty if no data has been gathered yet.)
```

jyavenard replied:
One more thing: I use the ability to know if there's an encode operation mid-flight or not to determine if it's okay to immediately write the audio frames to the container. 
Using the promise only for when flushing would lose that. 
Unless I called flush regularly. but those block the main thread for audio, so don't want to flush too often 

(in an earlier version I wasn't flushing the audio compressor when fetchData was called for this reason) I will take this as a follow up, making AudioSampleBufferCompressor::flush return a promise instead.

jyavenard replied:
Ok, I rewrote it , removing serialisation of decoding operations, I guess they could all happen in parallel and up to the encoder to serialise them as it already does.

It does rely however on the VideoEncoder to allow to call encode() before the previous promise resolved. I don't think it may always be the case.

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
```diff
@@ -0,0 +1,601 @@
+                m_lastVideoKeyframeTime = frame.second;
+            }
+            LOG(MediaStream, "encodePendingVideoFrames:encoding video frame:%f", frame.second.value());
+            return m_videoEncoder->encode({ WTFMove(frame.first), frame.second.microsecondsAs<int64_t>(), { } }, needVideoKeyframe);
```
It is a bit odd to do Ref{ m_currentOperation } but not protect m_videoEncoder here.

jyavenard replied:
m_videoEncoder isn't ref-counted.
protecting m_currentOperation is only there because the new safer CPP requires it (even though it makes no sense here)

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
```diff
@@ -0,0 +1,601 @@
+
+    m_lastAudioSampleTime = m_currentAudioSampleTime;
+    if (auto sampleBuffer = createAudioSampleBuffer(data, description, PAL::toCMTime(m_currentAudioSampleTime), sampleCount))
+        m_audioCompressor->addSampleBuffer(sampleBuffer.get());
```
Ref { m_audioCompressor } I guess...

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
```diff
@@ -0,0 +1,601 @@
+    if (m_audioCompressor)
+        RefPtr { m_audioCompressor }->finish();
+
+    flushPendingData()->whenSettled(MainThreadDispatcher::singleton(), [protectedThis = Ref { *this }, this] {
```
Why not weakThis here, it seems the only place where we take a strong ref?

jyavenard replied:
we're stopping the encoder, we want the completionHandler to be called under all circumstances. Keeping the ref ensure everything will be completed nicely and things fully cleaned up

youennf reviewed (commented):
Looks good to me overall. Some nits overall and either split part of the patch or add a test.

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm:
```diff
@@ -0,0 +1,589 @@
+        return m_videoEncoder->encode({ WTFMove(frame.first), frame.second.microsecondsAs<int64_t>(), { } }, needVideoKeyframe);
+    } };
+
+    VideoEncoder::EncodePromise::all(WTFMove(promises))->whenSettled(MainThreadDispatcher::singleton(), [weakThis = ThreadSafeWeakPtr { *this }, this](auto&&) {
```
Ce serait plus logique de faire m_pendingVideoEncode++ pour chaque appel m_videoEncoder->encode, et m_pendingVideoEncode-- dans chaque promise retournée.
Plus besoin de VideoEncoder::EncodePromise::all.

jyavenard replied:
That would need to dispatch twice as many tasks. The use of promise::all here is slightly more efficient

youennf reviewed (approved):
LGTM, please see comments below.

jyavenard committed:
[Cocoa] Add webm support to MediaRecorder
https://bugs.webkit.org/show_bug.cgi?id=281106
rdar://137560454

Reviewed by Youenn Fablet.

We add a MediaRecorderPrivateWriterWebM that uses the mkvmuxer from libwebm and the WebCodecs video encoders.
The WebM MediaRecorder is behind a preferences, currently disabled by default.

The MediaRecorderPrivateWriterWebM is currently used in the GPU process, in place of the MediaRecorderPrivateWriterCocoa
for ease of integration.
Performing the encoding and muxing will be moved to the Web Process in a follow-up change.

Amending existing tests to exercise the code.

* Configurations/WebKitProjectPaths.xcconfig:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-AV-audio-video-dataavailable-gpuprocess-expected.txt:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-AV-audio-video-dataavailable-gpuprocess.html:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-expected.txt:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate.html:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-requestData-expected.txt:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-requestData.html:
* LayoutTests/http/wpt/mediarecorder/mimeType-expected.txt:
* LayoutTests/http/wpt/mediarecorder/mimeType.html:
* LayoutTests/http/wpt/mediarecorder/pause-recording-expected.txt:
* LayoutTests/http/wpt/mediarecorder/pause-recording.html:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-mimetype-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-peerconnection.https-expected.txt:
* LayoutTests/platform/glib/TestExpectations: GStreamer MediaRecorder always create a MP4 one.
* Source/ThirdParty/libwebrtc/CMakeLists.txt:
* Source/ThirdParty/libwebrtc/Configurations/BaseTarget-libwebm.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/libwebm.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.exp:
* Source/ThirdParty/libwebrtc/Source/third_party/libwebm/CMakeLists.txt:
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/Modules/mediarecorder/MediaRecorder.cpp:
(WebCore::MediaRecorder::isTypeSupported):
* Source/WebCore/Modules/mediarecorder/MediaRecorderProvider.cpp:
(WebCore::MediaRecorderProvider::isSupported):
(WebCore::MediaRecorderProvider::isWebMAndSupported):
* Source/WebCore/Modules/mediarecorder/MediaRecorderProvider.h:
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::span8):
(WebCore::createOpusPrivateData):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.cpp:
(WebCore::MediaRecorderPrivateAVFImpl::create):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.cpp: Copied from Source/WebCore/Modules/mediarecorder/MediaRecorderProvider.h.
(WebCore::MediaRecorderPrivateWriter::create):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h: Added.
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::create):
(WebCore::AudioSampleBufferCompressor::AudioSampleBufferCompressor):
(WebCore::AudioSampleBufferCompressor::initAudioConverterForSourceFormatDescription):
(WebCore::AudioSampleBufferCompressor::getOutputSampleBuffer const):
(WebCore::AudioSampleBufferCompressor::isEmpty const):
(WebCore::AudioSampleBufferCompressor::getOutputSampleBuffer): Deleted.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm:
(WebCore::MediaRecorderPrivateWriter::create):
(WebCore::MediaRecorderPrivateWriterAVFObjC::create):
(WebCore::MediaRecorderPrivateWriterAVFObjC::compressedVideoOutputBufferCallback):
(WebCore::MediaRecorderPrivateWriterAVFObjC::compressedAudioOutputBufferCallback):
(WebCore::MediaRecorderPrivateWriterAVFObjC::MediaRecorderPrivateWriterAVFObjC):
(WebCore::MediaRecorderPrivateWriterAVFObjC::~MediaRecorderPrivateWriterAVFObjC):
(WebCore::MediaRecorderPrivateWriterAVFObjC::close):
(WebCore::MediaRecorderPrivateWriterAVFObjC::initialize):
(WebCore::MediaRecorderPrivateWriterAVFObjC::processNewCompressedVideoSampleBuffers):
(WebCore::MediaRecorderPrivateWriterAVFObjC::processNewCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateWriterAVFObjC::startAssetWriter):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendCompressedAudioSampleBufferIfPossible):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendCompressedVideoSampleBufferIfPossible):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendCompressedVideoSampleBuffer):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendCompressedSampleBuffers):
(WebCore::MediaRecorderPrivateWriterAVFObjC::flushCompressedSampleBuffers):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendVideoFrame):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendAudioSampleBuffer):
(WebCore::MediaRecorderPrivateWriterAVFObjC::finishedFlushingSamples):
(WebCore::MediaRecorderPrivateWriterAVFObjC::stopRecording):
(WebCore::MediaRecorderPrivateWriterAVFObjC::fetchData):
(WebCore::MediaRecorderPrivateWriterAVFObjC::completeFetchData):
(WebCore::MediaRecorderPrivateWriterAVFObjC::appendData):
(WebCore::MediaRecorderPrivateWriterAVFObjC::takeData):
(WebCore::MediaRecorderPrivateWriterAVFObjC::pause):
(WebCore::MediaRecorderPrivateWriterAVFObjC::resume):
(WebCore::MediaRecorderPrivateWriterAVFObjC::mimeType const):
(WebCore::MediaRecorderPrivateWriterAVFObjC::audioBitRate const):
(WebCore::MediaRecorderPrivateWriterAVFObjC::videoBitRate const):
(WebCore::MediaRecorderPrivateWriter::compressedVideoOutputBufferCallback): Deleted.
(WebCore::MediaRecorderPrivateWriter::compressedAudioOutputBufferCallback): Deleted.
(WebCore::MediaRecorderPrivateWriter::MediaRecorderPrivateWriter): Deleted.
(WebCore::MediaRecorderPrivateWriter::~MediaRecorderPrivateWriter): Deleted.
(WebCore::MediaRecorderPrivateWriter::close): Deleted.
(WebCore::MediaRecorderPrivateWriter::initialize): Deleted.
(WebCore::MediaRecorderPrivateWriter::processNewCompressedVideoSampleBuffers): Deleted.
(WebCore::MediaRecorderPrivateWriter::processNewCompressedAudioSampleBuffers): Deleted.
(WebCore::MediaRecorderPrivateWriter::startAssetWriter): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendCompressedAudioSampleBufferIfPossible): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendCompressedVideoSampleBufferIfPossible): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendCompressedVideoSampleBuffer): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendCompressedSampleBuffers): Deleted.
(WebCore::MediaRecorderPrivateWriter::flushCompressedSampleBuffers): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendVideoFrame): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendAudioSampleBuffer): Deleted.
(WebCore::MediaRecorderPrivateWriter::finishedFlushingSamples): Deleted.
(WebCore::MediaRecorderPrivateWriter::stopRecording): Deleted.
(WebCore::MediaRecorderPrivateWriter::fetchData): Deleted.
(WebCore::MediaRecorderPrivateWriter::completeFetchData): Deleted.
(WebCore::MediaRecorderPrivateWriter::appendData): Deleted.
(WebCore::MediaRecorderPrivateWriter::takeData): Deleted.
(WebCore::MediaRecorderPrivateWriter::pause): Deleted.
(WebCore::MediaRecorderPrivateWriter::resume): Deleted.
(WebCore::MediaRecorderPrivateWriter::mimeType const): Deleted.
(WebCore::MediaRecorderPrivateWriter::audioBitRate const): Deleted.
(WebCore::MediaRecorderPrivateWriter::videoBitRate const): Deleted.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.h: Copied from Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.h.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm: Added.
(WebCore::MediaRecorderPrivateWriterWebMDelegate::MediaRecorderPrivateWriterWebMDelegate):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addFrame):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::forceNewClusterOnNextFrame):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::finalize):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::hasAddedFrame const):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::protectedParent const):
(WebCore::MediaRecorderPrivateWriterWebM::create):
(WebCore::mkvCodeIcForMediaVideoCodecId):
(WebCore::codecStringForMediaVideoCodecId):
(WebCore::MediaRecorderPrivateWriterWebM::MediaRecorderPrivateWriterWebM):
(WebCore::MediaRecorderPrivateWriterWebM::~MediaRecorderPrivateWriterWebM):
(WebCore::MediaRecorderPrivateWriterWebM::close):
(WebCore::MediaRecorderPrivateWriterWebM::compressedAudioOutputBufferCallback):
(WebCore::MediaRecorderPrivateWriterWebM::initialize):
(WebCore::MediaRecorderPrivateWriterWebM::enqueueCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateWriterWebM::maybeStartWriting):
(WebCore::MediaRecorderPrivateWriterWebM::appendVideoFrame):
(WebCore::MediaRecorderPrivateWriterWebM::nextVideoFrameTime const):
(WebCore::MediaRecorderPrivateWriterWebM::resumeVideoTime const):
(WebCore::MediaRecorderPrivateWriterWebM::sampleTime const):
(WebCore::MediaRecorderPrivateWriterWebM::encodePendingVideoFrames):
(WebCore::MediaRecorderPrivateWriterWebM::appendAudioSampleBuffer):
(WebCore::MediaRecorderPrivateWriterWebM::flushEncodedQueues):
(WebCore::MediaRecorderPrivateWriterWebM::partiallyFlushEncodedQueues):
(WebCore::MediaRecorderPrivateWriterWebM::muxNextFrame):
(WebCore::MediaRecorderPrivateWriterWebM::stopRecording):
(WebCore::MediaRecorderPrivateWriterWebM::fetchData):
(WebCore::MediaRecorderPrivateWriterWebM::completeFetchData):
(WebCore::MediaRecorderPrivateWriterWebM::flushPendingData):
(WebCore::MediaRecorderPrivateWriterWebM::appendData):
(WebCore::MediaRecorderPrivateWriterWebM::takeData):
(WebCore::MediaRecorderPrivateWriterWebM::pause):
(WebCore::MediaRecorderPrivateWriterWebM::resume):
(WebCore::MediaRecorderPrivateWriterWebM::mimeType const):
(WebCore::MediaRecorderPrivateWriterWebM::audioBitRate const):
(WebCore::MediaRecorderPrivateWriterWebM::videoBitRate const):
(WebCore::MediaRecorderPrivateWriterWebM::maybeForceNewCluster):
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.cpp:
(WebKit::RemoteMediaRecorder::fetchData):
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.h:
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.messages.in:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionManagerProxy::seekableRangesVectorChanged):
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.cpp:
(WebKit::MediaRecorderPrivate::fetchData):

Canonical link: https://commits.webkit.org/285061@main
